PLAT-121331: Support Audio Guidance in ArcPicker, ArcSlider, FanSpeedControl, TemperatureControl, and WindDirectionControl

ybsung opened the pull request:
- `ArcPicker` and `ArcSlider` become spottable to support Audio Guidance when focused.
- If pressing `Up` or `Down` keys on them, the value in them increases or decreases.
So `Left` or `Right` keys should be pressed when navigating among those components.
- Then Audio Guidance in `ArcPicker`, `ArcSlider`, `FanSpeedControl`, `TemperatureControl`, and `WindDirectionControl` are supported properly.
- `AriaValuetextDecorator.js` in qa-a11y was added to decorate a `aria-valuetext` prop based on the value from `onChange`. It is used in ArcPicker, ArcSlider, FanSpeedControl, and WindDirectionControl views.
- The `horizontal` prop was added in the Arc family component's view `Section` so that the Arc components are shown horizontally and could be nevigable via `Right` or `Left` key because `Up` and `Down` keys are used to increase or decrease a value.
- `// eslint-disable-next-line jsx-a11y/role-has-required-aria-props` comment was added to work properly on webOS despite of the spec -`https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/role-has-required-aria-props.md`
- `ThemeContext` in `ThemeDecorator` is added to share `accent` and `highlight`.
- Rename qa-a11y Section's `vertical` prop as `horizontal`

Enact-DCO-1.0-Signed-off-by: YB Sung (yb.sung@lge.com)

MikyungKim commented:
@ybsung Please resolve the merge conflict.

MikyungKim commented on ArcPicker/ArcPicker.js:
```diff
@@ -222,6 +253,7 @@ const ArcPickerBase = kind({
 const ArcPickerDecorator = compose(
 	Changeable,
 	ArcPickerBehaviorDecorator,
+	Spottable,
```
![image](https://user-images.githubusercontent.com/5037429/101303173-cced3880-3880-11eb-99c6-bac00e0defb9.png)
Is there a way to only focus on an arc segment or not? It looks so ugly...

ybsung replied:
I fixed as you mentioned so that the Arc could get the accent color instead of ArcPicker itself.

MikyungKim commented on ArcPicker/ArcPicker.js:
```diff
@@ -155,6 +169,23 @@ const ArcPickerBase = kind({
+
+			if (!disabled) {
+				if (isDown(ev.keyCode)) {
+					onClick(children[Math.max(index - 1, 0)])(ev);
```
Should we use `onClick` here? how about `onChange` like `ArcSlider`?

ybsung replied:
As we discussed, let's leave it as-is because `onClick` is defined as public now.

MikyungKim commented on CHANGELOG.md:
```diff
@@ -14,7 +14,15 @@ The following is a curated list of changes in the Enact agate module, newest cha
+
+### Changed
+
+- `agate/ArcPicker` to read out audio guidance when focused via 5-way keys
+- `agate/ArcSlider` to read out audio guidance when focused via 5-way keys
+- `agate/FanSpeedControl` to read out audio guidance when focused via 5-way keys
+- `agate/TemperatureControl` to read out audio guidance when focused via 5-way keys
+- `agate/WindDirectionControl` to read out audio guidance when focused via 5-way keys
```
Why did you repeat the same log? you could do like
`agate/ArcPicker`, `agate/ArcSlider`, ... in one log.

MikyungKim commented on ArcPicker/ArcPicker.js:
```diff
@@ -155,6 +169,23 @@ const ArcPickerBase = kind({
 
+	handlers: {
+		onKeyDown: (ev, props) => {
+			const {children, disabled, onClick, value} = props;
```
Can we destruct these from above?

MikyungKim commented on ArcPicker/ArcPicker.module.less:
```diff
@@ -24,5 +24,14 @@
 			height: @agate-arc-picker-value-display-height;
 		}
+
+		.spottable({
+			&:focus {
```
I think we have `.focus` mixin for this.

MikyungKim commented on CHANGELOG.md:
```diff
@@ -14,7 +14,11 @@ The following is a curated list of changes in the Enact agate module, newest cha
 - `agate/TimePicker` prop `onSpotlightDisappear` and `spotlightDisabled`
- 
+
+### Changed
```
Please add `Added` for the added props.

MikyungKim commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ The following is a curated list of changes in the Enact agate module, newest cha
 
 ### Added
 
+- `agate/ArcPicker`, `agate/ArcSlider` prop `accentColor` to customize accent color when focused
```
```suggestion
- `agate/ArcPicker` and `agate/ArcSlider` prop `accentColor` to customize accent color when focused
```

MikyungKim commented on CHANGELOG.md:
```diff
@@ -14,7 +15,11 @@ The following is a curated list of changes in the Enact agate module, newest cha
+
+### Changed
+
+- `agate/ArcPicker`, `agate/ArcSlider`, `agate/FanSpeedControl`, `agate/TemperatureControl` and `agate/WindDirectionControl` to read out audio guidance when focused via 5-way keys
```
```suggestion
- `agate/ArcPicker`, `agate/ArcSlider`, `agate/FanSpeedControl`, `agate/TemperatureControl`, and `agate/WindDirectionControl` to read out audio guidance when focused via 5-way keys
```

MikyungKim commented on samples/sampler/stories/default/ArcPicker.js:
```diff
@@ -23,6 +23,8 @@ storiesOf('Agate', module)
 			return (
 				<ArcPicker
+					accentColor={select('accentColor', prop.colors, Config)}
+					backgroundColor={select('backgroundColor', prop.colors, Config)}
```
?

MikyungKim reviewed (approved):
LGTM

seunghoh reviewed (changes requested):
The focused accent color is purple even on "silicon" skin. Please update with regard to the associated color set

ybsung commented:
> The focused accent color is purple even on "silicon" skin. Please update with regard to the associated color set

Hello @seunghoh 

The `foregroundColor` and `backgroundColor` props also exists in ArcPicker and ArcSlider. In my opinion, it is inconsistency to update only `accentColor` prop.

If I have to update all of those props, our Agate APIs will be broken due to removing the `foregroundColor` and `backgroundColor` props. In addition, the `foregroundColor` changes. For example, the TemperatureControl uses '#007aff' 
 and '#f24949' depending on the temperature (cold or hot).

But if it would be better to go ahead, please let me know. I'll update my PR.

seunghoh reviewed (changes requested):
ui-test fails like 
AssertionError in "ArcPicker default should select the third arc when it is clicked"
expected 'rgb(139,126,254)' to equal 'rgb(68,68,68)'

Please update ui-test with regard to the ux/gui changes

seunghoh reviewed (changes requested):
There are several things to be fixed
1. "Deactivated" is not read focused on disabled control
2. NAN error on click the slider control
3. TemperatureControl read value should be the same as shown number + unit

ybsung commented:
> The focused accent color is purple even on "silicon" skin. Please update with regard to the associated color set

: Fixed. ThemeDecorator passes `accent` and `hightlight` variables as a context in this PR. So I used it in ArcPicker and ArcSlider when focused.

> ui-test fails like
> AssertionError in "ArcPicker default should select the third arc when it is clicked"
> expected 'rgb(139,126,254)' to equal 'rgb(68,68,68)'
> 
> Please update ui-test with regard to the ux/gui changes

: Fixed after refreshing the branch based on the latest `develop` branch

> There are several things to be fixed
> 
> 1. "Deactivated" is not read focused on disabled control


: Not yet

> 2. NAN error on click the slider control

: Fixed.

> 3. TemperatureControl read value should be the same as shown number + unit

: Fixed by passing `aria-valuetext` prop

In addtion, qa-a11y Section's `vertical` prop is renamed as `horizontal` based on the review.

seunghoh commented on ThemeDecorator/ThemeDecorator.js:
```diff
@@ -280,10 +287,10 @@ const ThemeDecorator = hoc(defaultConfig, (config, Wrapped) => {
 
 			return (
-				<React.Fragment>
+				<ThemeContext.Provider value={{accent, highlight}}>
```
accent, highlight should be skin's default color, not fixed color if app developer doesn't set

ybsung replied:
Depending on the current skin, I set the default accent and highlight color.

For other components, those color are defined in less files. But `ArcPicker` and `ArcSlider` uses SVG format, then can't use them in the less files. So we have to define them in JavaScript files - ThemeDecorator.js separately. That's the reason why I have to define the `defaultColors` variable in it.

seunghoh commented on samples/qa-a11y/src/components/Section.js:
```diff
@@ -22,20 +23,20 @@ const SectionBase = kind({
 			<Cell size={1500} {...rest}>
 				<Heading showLine>{title}</Heading>
-				<Row className={css.componentDemo}>
+				<Row className={classNames(css.componentDemo, css.vertical)}>
```
css.vertical or css.horizontal?

ybsung replied:
Fixed.

seunghoh commented on CHANGELOG.md:
```diff
@@ -6,15 +6,22 @@ The following is a curated list of changes in the Enact agate module, newest cha
 ### Added
 
+- `agate/ArcPicker` and `agate/ArcSlider` prop `disabled`
+- `agate/ArcSlider` prop `aria-valuetext`
```
What's the purpose of this new added prop? 
for example, `agate/ArcPicker` and `agate/ArcSlider` prop `disabled` to support blah blah...

ybsung replied:
Updated CHANGELOG.md.

seunghoh commented on CHANGELOG.md:
```diff
@@ -6,15 +6,22 @@ The following is a curated list of changes in the Enact agate module, newest cha
 - `agate/Marquee` component
 - `agate/Picker` prop `noAnimation` and `wrap`
 - `agate/RangePicker` prop `noAnimation` and `wrap`
+- `agate/ThemeDecorator` context `ThemeContext`
```
Same here, what's the purpose?

ybsung replied:
Updated CHANGELOG.md.

MikyungKim commented on samples/qa-a11y/src/components/AriaValuetextDecorator.js:
```diff
@@ -0,0 +1,37 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+
+const AriaValuetextDecorator  = hoc((config, Wrapped) => {
```
`AriaValueTextDecorator`?

MikyungKim commented on CHANGELOG.md:
```diff
@@ -6,15 +6,22 @@ The following is a curated list of changes in the Enact agate module, newest cha
 - `agate/Marquee` component
 - `agate/Picker` prop `noAnimation` and `wrap`
 - `agate/RangePicker` prop `noAnimation` and `wrap`
+- `agate/ThemeDecorator` context `ThemeContext` to share the context including `accent` and `highlight` properties
```
This is just "what" you did. I think we need to add "why" this change needed here.

MikyungKim commented on ThemeDecorator/ThemeDecorator.js:
```diff
@@ -280,10 +323,10 @@ const ThemeDecorator = hoc(defaultConfig, (config, Wrapped) => {
 					{customSkin ? <CustomizableSkinStyle className={customizableSkinClassName} accent={accent} highlight={highlight} /> : null}
-					<App {...rest} className={allClassNames} />
-				</React.Fragment>
+					<App {...rest} accent={accent} className={allClassNames} />
```
Don't we need `highlight` also here?

ybsung commented:
There were unit test fails related `ThemeContext`.
If a component with the `ThemeContext` context isn't wrapped with `ThemeDecorator` like unit tests,
the context of `ThemeContext` is `undefined`.
I didn't consider this case before. So I handled it additionally.
